src/HonzaBotner.Discord.Services/Reactions/PinHandler.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using DSharpPlus.Entities;
+using DSharpPlus.EventArgs;
+using HonzaBotner.Discord.Services.Options;
+using HonzaBotner.Services.Contract;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace HonzaBotner.Discord.Services.Reactions
+{
+    public class PinHandler : IReactionHandler
+    {
+        private readonly PinOptions _pinOptions;
+        private readonly DiscordWrapper _discordWrapper;
+        private readonly ILogger<PinHandler> _logger;
+
+        public PinHandler(IOptions<PinOptions> pinOptions, DiscordWrapper discordWrapper, ILogger<PinHandler> logger)
+        {
+            _pinOptions = pinOptions.Value;
+            _discordWrapper = discordWrapper;
+            _logger = logger;
+        }
+
+        public async Task<IReactionHandler.Result> HandleAddAsync(MessageReactionAddEventArgs eventArgs)
+        {
+            if (eventArgs.Message.Pinned) return IReactionHandler.Result.Continue;
+
+            try
+            {
+                DiscordEmoji pinEmoji = DiscordEmoji.FromName(_discordWrapper.Client, _pinOptions.EmojiName);
+
+                if (eventArgs.Emoji.Equals(pinEmoji))
+                {
+                    var reactions = await eventArgs.Message.GetReactionsAsync(pinEmoji);
+
+                    if (reactions.Count >= _pinOptions.Treshold)
+                    {
+                        await eventArgs.Message.PinAsync();
+                    }
+                    else
+                    {
+                        Dictionary<ulong, int> roleToScore = new Dictionary<ulong, int>();
+                        foreach (var keyValuePair in _pinOptions.RoleToWeightMapping)
+                        {
+                            try
+                            {
+                                ulong roleId = UInt64.Parse(keyValuePair.Key);
```

Comment:
```suggestion
                                ulong roleId = ulong.Parse(keyValuePair.Key);
```
Also consider using ulong.TryParse